this causes `storage.TestStoreRangeDownReplicate` to fail with:
```
panic: snapshot is temporarily unavailable

goroutine 6274 [running]:
panic(0x16f1120, 0xc8201e40c0)
	/home/marc/go/src/runtime/panic.go:464 +0x3e6
github.com/coreos/etcd/raft.(*raftLog).term(0xc820d3c150, 0xf, 0x2, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/log.go:238 +0x1a9
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc8205c8ea0, 0x2)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/raft.go:286 +0xc4
github.com/coreos/etcd/raft.(*raft).bcastAppend(0xc8205c8ea0)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/raft.go:360 +0xac
github.com/coreos/etcd/raft.stepLeader(0xc8205c8ea0, 0x2, 0x0, 0x1, 0x0, 0x0, 0x0, 0xc820d8cc80, 0x1, 0x1, ...)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/raft.go:622 +0x560
github.com/coreos/etcd/raft.(*raft).Step(0xc8205c8ea0, 0x2, 0x0, 0x1, 0x0, 0x0, 0x0, 0xc820d8cc80, 0x1, 0x1, ...)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/raft.go:579 +0x66d
github.com/coreos/etcd/raft.(*RawNode).Propose(0xc82090ac40, 0xc820daa380, 0x66, 0x66, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/coreos/etcd/raft/rawnode.go:132 +0x121
github.com/cockroachdb/cockroach/storage.defaultProposeRaftCommandLocked(0xc820705500, 0xc820d8a840, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/storage/replica.go:1422 +0x385
github.com/cockroachdb/cockroach/storage.(*Replica).proposePendingCmdLocked(0xc820705500, 0xc820d8a840, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/storage/replica.go:1380 +0x6e
github.com/cockroachdb/cockroach/storage.(*Replica).proposeRaftCommand(0xc820705500, 0x7f870eb831a8, 0xc82096e3c0, 0x0, 0xb6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/storage/replica.go:1367 +0x6d7
github.com/cockroachdb/cockroach/storage.(*Replica).addWriteCmd(0xc820705500, 0x7f870eb831a8, 0xc82096e3c0, 0x0, 0xb6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/storage/replica.go:1286 +0x333
github.com/cockroachdb/cockroach/storage.(*intentResolver).processIntentsAsync.func2()
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/storage/intent_resolver.go:335 +0x683
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunLimitedAsyncTask.func1(0xc820902a80, 0x2146129, 0x1a, 0x152, 0xc820803320, 0xc8207e5740)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:218 +0x8d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunLimitedAsyncTask
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:219 +0x1e2
exit status 2
```

--- a/storage/helpers_test.go
+++ b/storage/helpers_test.go
@@ -95,13 +95,14 @@ func (s *Store) LogReplicaChangeTest(txn *client.Txn, changeType roachpb.Replica
 // to be held and it will block instead of returning
 // ErrSnapshotTemporaryUnavailable.
 func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
 	for {
+		r.mu.Lock()
 		snap, err := r.Snapshot()
+		snapshotChan := r.mu.snapshotChan
+		r.mu.Unlock()
 		if err == raft.ErrSnapshotTemporarilyUnavailable {
 			var ok bool
-			snap, ok = <-r.mu.snapshotChan
+			snap, ok = <-snapshotChan
 			if ok {
 				return snap, nil
 			}

--- a/storage/helpers_test.go
+++ b/storage/helpers_test.go
@@ -22,8 +22,11 @@
 package storage
 
 import (
+	"time"
+
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util/retry"
 	"github.com/coreos/etcd/raft"
 	"github.com/coreos/etcd/raft/raftpb"
 )
@@ -95,12 +98,22 @@ func (s *Store) LogReplicaChangeTest(txn *client.Txn, changeType roachpb.Replica
 // to be held and it will block instead of returning
 // ErrSnapshotTemporaryUnavailable.
 func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
-	for {
+	retryOptions := retry.Options{
+		InitialBackoff: 1 * time.Millisecond,
+		MaxBackoff:     50 * time.Millisecond,
+		Multiplier:     2,
+	}
+	for retry := retry.Start(retryOptions); retry.Next(); {
 		r.mu.Lock()
 		snap, err := r.Snapshot()
 		snapshotChan := r.mu.snapshotChan
 		r.mu.Unlock()
 		if err == raft.ErrSnapshotTemporarilyUnavailable {
+			if snapshotChan == nil {
+				// The call to Snapshot() didn't start an async process due to
+				// rate limiting. Try again later.
+				continue
+			}
 			var ok bool
 			snap, ok = <-snapshotChan
 			if ok {
@@ -113,4 +126,5 @@ func (r *Replica) GetSnapshot() (raftpb.Snapshot, error) {
 			return snap, err
 		}
 	}
+	panic("unreachable") // due to infinite retries
 }

--- a/storage/replica_raftstorage.go
+++ b/storage/replica_raftstorage.go
@@ -381,6 +381,11 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 		}
 	}
 
+	// See if there is already a snapshot running for this store.
+	if !r.store.AcquireRaftSnapshot() {
+		return raftpb.Snapshot{}, raft.ErrSnapshotTemporarilyUnavailable
+	}
+
 	// Use an unbuffered channel so the worker stays alive until someone
 	// reads from the channel, and can abandon the snapshot if it gets
 	// stale.
@@ -389,6 +394,7 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 		defer close(ch)
 		snap := r.store.NewSnapshot()
 		defer snap.Close()
+		defer r.store.ReleaseRaftSnapshot()
 		// Delegate to a static function to make sure that we do not depend
 		// on any indirect calls to r.store.Engine() (or other in-memory
 		// state of the Replica). Everything must come from the snapshot.
@@ -405,6 +411,8 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 		}
 	}) {
 		r.mu.snapshotChan = ch
+	} else {
+		r.store.ReleaseRaftSnapshot()
 	}
 
 	if r.store.ctx.BlockingSnapshotDuration > 0 {

--- a/storage/store.go
+++ b/storage/store.go
@@ -288,6 +288,11 @@ type Store struct {
 	initComplete            sync.WaitGroup // Signaled by async init tasks
 	raftRequestChan         chan *RaftMessageRequest
 
+	// This is 1 if there is an active raft snapshot. This field must be checked
+	// and set atomically.
+	// TODO(marc): This may be better inside of `mu`, but is not currently feasible.
+	hasActiveRaftSnapshot int32
+
 	// drainLeadership holds a bool which indicates whether Replicas should be
 	// allowed to acquire or extend leader leases; see DrainLeadership().
 	//
@@ -1594,6 +1599,17 @@ func (s *Store) processRangeDescriptorUpdateLocked(rng *Replica) error {
 // NewSnapshot creates a new snapshot engine.
 func (s *Store) NewSnapshot() engine.Engine {
 	return s.engine.NewSnapshot()
+}
+
+// AcquireRaftSnapshot returns true if a new raft snapshot can start.
+// If true is returned, the caller MUST call ReleaseRaftSnapshot.
+func (s *Store) AcquireRaftSnapshot() bool {
+	return atomic.CompareAndSwapInt32(&s.hasActiveRaftSnapshot, 0, 1)
+}
+
+// ReleaseRaftSnapshot decrements the count of active snapshots.
+func (s *Store) ReleaseRaftSnapshot() {
+	atomic.SwapInt32(&s.hasActiveRaftSnapshot, 0)
 }
 
 // Attrs returns the attributes of the underlying store.

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -1962,3 +1962,24 @@ func TestStoreChangeFrozen(t *testing.T) {
 		}
 	}
 }
+
+func TestStoreNoConcurrentRaftSnapshots(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	store, _, stopper := createTestStore(t)
+	defer stopper.Stop()
+
+	if !store.AcquireRaftSnapshot() {
+		t.Fatal("expected true")
+	}
+
+	if store.AcquireRaftSnapshot() {
+		t.Fatalf("expected false")
+	}
+
+	store.ReleaseRaftSnapshot()
+
+	if !store.AcquireRaftSnapshot() {
+		t.Fatal("expected true")
+	}
+	store.ReleaseRaftSnapshot()
+}

